### ISSUE TITLE
Default ODM mapper for MongoDB?

### ISSUE THREAD

**Issue by docsolver** [yiisoft/yii2#4899]
The documentation of the `yii2-mongodb` extension states:

> This extension does not provide any special way to work with embedded documents (sub-documents). General recommendation is avoiding it if possible

I understand how including an Object Document Mapper for embedded documents would be very complex, but to me its absence defeats the whole purpose of using MongoDB with ActiveRecord. After all, due to its NoSQL-nature, MongoDB relies heavily on embedded documents as an alternative to relational joins. I was very excited to see MongoDB support out-of-the-box in Yii2, but I believe that this omission will severely hinder Yii2+MongoDB adoption for most people.

Instead of building a native solution, there are some existing tools that might be plugged in to do this job. It seems that the most important ones are [Doctrine Mongodb ODM](http://www.doctrine-project.org/projects/mongodb-odm.html) and [Mandango](http://mandango.org/).

I am eager to adopt MongoDB, since it would solve some of the constraints I am facing with a relational approach. Unfortunately my experience with MongoDB and hacking the internals of Yii2 is insufficient to build a decent solution myself.

However, I assume that I am not the only one facing this problem, so maybe someone else has already successfully combined one of these tools (or something similar) with Yii2 ActiveRecord to enable MongoDB-based ActiveRecord models with embedded models as subdocuments? If so, maybe this approach could be included in the `yii2-mongodb` extension or in the documentation?

**Comment by alexandernst** [comment]
@cebe I'm considering spending some time on this and implement it, but I'm not really sure where to start nor how exactly to do it. Do you have any guide/plans for this that you could share with me?

**Comment by cebe** [comment]
Well, for elasticsearch all sub documents are already represented as PHP arrays because the communication is in json and we simply json_encode/decode the values so there is nothing to be implemented additionally. For mongodb I have no idea what has to be done to implement this. How is the data of sub-documents retrieved currently?

**Comment by klimov-paul** [comment]
For the MongoDB is it the same.

**Comment by Sammaye** [comment]
One thing that is being forgotten her is validation, manipulation, and most importantly (for performance) population of said subdocuments.

All of these points, population being most significant, is one reason why I dropped making this myself and made a pesonalised suboducment loader specifically designed for my needs.

**Comment by alexandernst** [comment]
@Sammaye Care to explain a little further the problem? (maybe with actual pseudo-code examples). Maybe we all can get to a solution (or throw away this entire bug if a solution is not possible).

**Comment by Sammaye** [comment]
The first problem I found was the difference of validating each model and the array as whole between objectual arrays and value based arays.

For example:

    [1,2,3,4,5]

Should have the of of `int` applied to each element but `max => 10` to the entire array.

and:
```javascript
{
    i: 'fgh',
    f: 1,
    g: [1,2,3,4,5]
}
```
Should have `string` and `int` applied and the nested document and then validated as a whole to ensure non-legit attributes are not entered.

I found that to validate `[1,2,34,5]` I required to hack Yii2 in an extremely nasty way add a model with only one field `value` which was used to apply the model attributes and attribute cleaning.

I foun I also needed to copy a lot of the `Model` to the suboducment moel sncer having `IteratorAggregate` makes things really hard especially if you want to create seamless iteration by implementing `Iterator` (which you cannot).

When it came to populating peformance is a killer. Some suboducments could have 100's of rows, I already know that you cannot house that many moels in memory (I regularly deal with out of memory errors in Yii2 on about 300meg per PHP process when I try to do too much in one go) so then  you must make lazy loaded suboducments but then a problem arose there which I have since forgotten.

One of the biggest problems was that after each time I coded the subdocumnt model class to do one thing I realised hat I also wanted to use it on another scenario too and whe that scenario was added I realied that it woul never scale to even the smallest of subdocuments.

For example, string keys for arrays, i.e. sessions:
```javascript
{
    '4456ghdtyffg0988' => {}
    '454rfghrfhdc?bnvghu7y' => {}
}
```
which I use very successfully for certain parts of session handling without having to draw MongoDB into in-memory operations which would be induced by the more standard approach.

That should get you started a little. I am sure I have more but that's enough for the min

**Comment by cebe** [comment]
@Sammaye so what you want to have here is to populate models from array valued attributes in the record? How would you define which model needs to be populated for which attributes? Maybe you could show your code you came up with so we can see how it works.

**Comment by Sammaye** [comment]
I don't actually have the code I had, it was a couple of months ago and after I failed to produce something that could perfoormantly add suboducment in a blanket fashion I gave up and in anger deleted the code. 

The one I have now is more simple of the sort:

```php
     public function validateArraySubdocument($attribute, $params)
     {
     	$errors = false;
     	$value = $this->$attribute;
     	
     	if(is_string($value)){
     		$value = explode(',', $value);
     	}
     	
     	$newValue = [];

     	if(isset($params['required']) && $params['required'] && empty($value)){
     		$this->addError(
     			$attribute, 
     			Yii::t(
     				'yii', 
     				'{attribute} cannot be empty.', 
     				['attribute' => $this->getAttributeLabel($attribute)]
				)
			);
     		return false;
     	}elseif(empty($value)){
     		return true;
     	}
     	
     	if(isset($params['min']) && $params['min'] > count($value)){
     		$this->addError(
     			$attribute, 
     			Yii::t(
     				'yii', 
     				'{attribute} has too few entries.', 
     				['attribute' => $this->getAttributeLabel($attribute)]
				)
			);
     		return false;
     	}
     	
     	if(isset($params['max']) && $params['max'] < count($value)){
     		$this->addError(
     			$attribute, 
     			Yii::t(
     				'yii', 
     				'{attribute} has too many entries.', 
     				['attribute' => $this->getAttributeLabel($attribute)]
				)
			);
     		return false;
     	}
     	
     	$vModel = new DynamicModel();
     	//$vModel = new ValueModel();
     	$vModel->rules(isset($params['rules']) ? $params['rules'] : []);

     	foreach($value as $k => $v){
     		
     		$vModel->clearAttributes();
     		$vModel->clearErrors();
     		
     		$vModel->setAttributes(['value' => $v]);
     		$vModel->validate();
     		
     		if($vModel->getErrors()){
     			$this->addError(
     				$attribute, 
     				Yii::t(
     					'yii', 
     					'{attribute} was not filled in correctly.', 
     					['attribute' => $this->getAttributeLabel($attribute)]
					)
				);
     			$errors = true;
     			break;
     		}
     		
     		$newValue[] = $vModel->value;
     	}
     	
     	if($errors){
     		return false;
     	}
     	
     	$this->$attribute = $newValue;
     	return true;
     }
```

and it only works for my specific uses, powerd by (in th `rules()` function):
```php
			[
				'languages',
				'validateArraySubdocument',
				'params' => [
					'max' => 10,
					'rules' =>
						[
							['$', 'common\components\IntRangeValidator', 'range' => array_keys($this->languages())],
						]
				]
			],
			
			[
				'personality',
				'validateArraySubdocument',
				'params' => [
					'max' => 10,
					'rules' =>
						[
							['$', 'common\components\IntRangeValidator', 'range' => array_keys($this->personality())],
						]
				]
			],
			 
			[
				'education',
				'validateArraySubdocument',
				'params' => [
					'max' => 5,
					'rules' =>
						[
							['$', 'common\components\IntRangeValidator', 'range' => array_keys($this->education())],
						]
				]
			],
			 
			[
				'pets',
				'validateArraySubdocument',
				'params' => [
					'max' => 5,
					'rules' =>
						[
							['$', 'common\components\IntRangeValidator', 'range' => array_keys($this->pets())],
						]
				]
			],
```

All these suboducments represent non model arrays, so here is no defined model and the subdocument is like:

    [1,2,3,4,5]

respresenting IDs of multiple select boxes.

However, before when I was coding a dynamic functional extension for this: in order to understand the structure and population rules of the suboducment I added a couple of fields like so:
```php
[
    'attribute' => 'c',
    'type' => manyArray|manyObject|oneArray|oneObject,
    'model' => 'Test',
    'rules' => []
]
```
whereby the type would switch the functionality of what was required. The `model` field was not required but if it was there the suboducment extension would not use `DynamicModel` to produce a validatable model for the part sof a subdocument.

And within the rules ou would have:
```php
[
   [ '$',  //rule ]
   [ 'd', //rule ]
]
```
`$` would represent validation functions for the suboducment as a whole whereas `d` would be a rule added to the model if the model field were not defined.

That is essentially how it used to juge cetain things about data assignment.

Populating the subodcument posed it's own problems in turn.

**Comment by Sammaye** [comment]
I think my biggest fault was that I tried to make `manyArray|manyObject|oneArray|oneObject` all one class for one.

**Comment by Sammaye** [comment]
As an addition I should mention that the level of error reporting for suboducments still exists, even for my current code, as to saying whether or not they validated. Html helper would need some rewriting to it to working. I still haven't done that yet, however, I might.

You may also want to make accessors for errors on suboducments for the `Model` like:
```php
public function getSubodcumentError()
{
   // some fun stuff here
}
```
but more elegant since that ain't pretty.

**Comment by alexandernst** [comment]
Sorry for leaving this unattended for so long;
So, from what I understand, there are 3 major problems:

* manipulation
* validation
* population

For the sake of this example, I'll use this model/data structure:

    _id: '3456345b6serbts34sdfbtsdf',
    type: [
        "admin",
        "client"
    ],
    personal_info: {
        name: 'Bob',
        surnames: ['', '', '', ''....],
        bday: ISODate(....),
    },
    credentials: {
        email: 'qqq@www.com',
        pass: '123123',
        salt: '2345c2345v345wq'
    },

### Manipulation

This one seems pretty easy to me. Currently, modifying a sub-document is possible only as a hole. See:

    $person = Person::findOne(["_id": "3456345b6serbts34sdfbtsdf"]);

    $person->credentials = [
        //new values here
    ];  <---- this runs fine

    $person->credentials->password = "new password"; <--- this will fail, yii2/base/InlineAction.php:55

IMHO manipulation should be done just as I tried in the last line. And that can be done easily, unless I'm missing something. If so, please correct me.

### Validation

This one shouldn't be hard either.

We can currently declare quite some rules for Models. See:

    public $username;
    public $email;
    public $subject;
    public $body;

    public function rules() {
        return [
            // username, email and password are required in "register" scenario
            [['username', 'email', 'password'], 'required', 'on' => 'register'],
    
            // username should be a string (>= 3 and <= 12)
            ['username', 'string', 'min' => 3, 'max' => 12],
        ];
    }

I can't really see any problem in extending this a little bit further for MongoDB:

    public $_id;
    public $type = [];
    public $personal_info = ['name', 'surnames', 'bdate'];
    public $credentials = ['email', 'pass', 'salt'];

    public function rules() {
        return [
            // email and password are required in "register" scenario
            [['credentials.email', credentials.pass'], 'required', 'on' => 'register'],
    
            // username should be a string (>= 3 and <= 12)
            ['personal_info.name', 'string', 'min' => 3, 'max' => 12],
            ['personal_info.bdate', 'ISODate'],
            ['personal_info.surnames', 'ArrayofStrings']
        ];
    }

The `ArrayOfStrinigs` could be, of course, one of many types, such as: `ArrayOfInts`, `ArrayOfArrays`, etc... Actually, there won't be too many, as you can't assign infinite types of data to a JSON key. You can actually assign just the following ones, so we'd have to implement only 6 different types:

* string
* integer
* array
* object
* bool
* null

As per docs: http://interactive-matter.org/wp-content/uploads/2010/08/JSON-500x457.png

### Population

This one is indeed a little bit harder in terms of code, but not impossible. The main point here is performance, which is fixable with lazy loading. IMHO lazy loading should be done on all sub-documents, which on the other hand can be easily detected as we already know which values are sub-documents and which values are plain fields (see my previous point, `Validation`).

That said, a pseudo-code of this would look like:

    $person = Person::findAll()->limit(50);
    // At this point the MongoDB extension should detect which attributes of Person are plain fields
    // and which ones are sub-documents, and retrieve only the ones that are plain fields.

    $person->credentials->pass = "foo bar";
    // We're trying to access `credentials`, which got defined as an array, which means that it's a
    // sub-document, which means that MongoDB extension should retrieve the `credentials` field
    // for the model with id `$person->_id` and assign that value to the current model.

    $person->save();
    // This should check the rules we defined in `Validation` and add errors if any.
    // IMHO easily done as we can iterate all the values and all the keys and check for the type
    //  we specified in `Validation`.

What do you think @Sammaye @cebe @klimov-paul @tadaszelvys  ?

**Comment by Sammaye** [comment]
What about many nested objects and the use of subclasses with them?

**Comment by Sammaye** [comment]
I never actually use this setup:
```javascript
personal_info: {
    name: 'Bob',
    surnames: ['', '', '', ''....],
    bday: ISODate(....),
},
```
Since I find there is no advantage, not even in coding since you normally have to pick out the properties and validate them anyway, so this pattern is not one regularly used by experienced MongoDB users.

**Comment by alexandernst** [comment]
@Sammaye You'd define/use them the same way.

Structure:

    _id: '',
    personal_info: [
        name: '',
        surnames: []
    ],
    history: [
        logins: [
            {
                'date': ISODate(),
                'ip': x.y.z.a
            },
            {
                'date': ISODate(),
                'ip': x.y.z.a
            }
        ],
        purchases: {
            'online': [
                {
                    'date': ISODate(),
                    'amount': n
                },
                {
                    'date': ISODate(),
                    'amount': n
                }
            ],
            'physical': [
                {
                    'date': ISODate(),
                    'amount': n
                },
                {
                    'date': ISODate(),
                    'amount': n
                }
            ]
        }
    ]

### Manipulation

    $person = Perso::find();
    $person->history->purchases->online[1]->['amount'];

### Validation

    public $history = [
        'logins',
        'purchases' => [
            'online',
            'physical'
        ]
    ];

    public function rules() {
        return [
            ['history', 'Array'],
            ['history.logins', 'ArrayOfObjects'],
            ['history.logins.date', 'ISODate'],
            ['history.logins.ip', 'IPValidator'],
            ['history.purchases', 'Object'],
            ['history.purchases.online', 'ArrayOfObjects'],
            ['history.purchases.date', 'ISODate'],
            ['history.purchases.amount', 'Integer'],
            ['history.purchases.physical', 'ArrayOfObjects'],
            ...
        ];
    }

### Population

Same principle I described in my previous comment.

Please note that <b>this</b> concrete example is, well, just an example. It doesn't represent anything more than a structure filled with dummy values so we can wrap our brains on top of that idea and get this issue to something. It certainly doesn't represent any regular patterns for data storage.

**Comment by klimov-paul** [comment]
In general I whould call this feature 'sub model' or 'nested model'. It is better to generic - not bound to MongoDB. So it can be used for regular models as well as for ElasticSearch for instance.
The goal is creating a mechanism, which expand particular attribute of the parent model into a sub model. So we need some sort of mapping: 
```
app\models\Person->history => app\models\History
```
The most common question in general: in which syntax particular model shold define sub models?
Note: each defined sub-model property should have an ability to be get and set. So both following statements should work:
```php
$person = Person::findOne();
echo $person->history->date;

$person = new Person();
$history = new PersonHistory();
$history->date = strtotime('NOW');
$person->hitory = $history;
```

1) Manipulation

There is a problem: if we expand particular model attribute into another model instance, there is a chance developer may actually need a pure attribute value at some point. There should be an opportunity to manipulate pure complex attribute value as array as well as a sub model.
This produces following problems:
- Complex attribute and sub model attribute can not have the same name. If by invocation of `$person->history` we receive model instance, the pure array attribute should have other name like `$person->historyData` or naming can be in vice versa: `$person->historyModel` and `$person->history`.
- There could be a conflict between pure complex attribute data and data of the sub-model. Which should take precedence?

2) Validation

I can not see much problem here. Sinse each sub-model is a model, if you need to validate it you simply call related method:
```php
$person->history->validate();
```
The main question is should the owner model trigger such validation automatically, while performing own `validate()`? If it does so it will load (instantiate) all sub-models in order to validate them - no matter if it is needed or not.
I suppose, this can be solved in easy way: we can create a special Validator class called 'SubModelValidator', which if applied will validate sub-model matching specified attribute, so main model validation rules can look following:
```php
public function rules()
{
    return [
        ['history', 'subModel']
    ];
}
```
Such rules can vary by scenario or simply omitted if needed.

3) Population

I am afraid it is not a good point to chase ability of populating nested models via owner `populate()` invocation. Just as well as populating models related via `hasOne()` or `hasMany()`. So I suppose it will be simple: you need to pouplate sub-model manually if you are expect its data preset in request:
```php
$person = Person::findOne();
$person->populate($request->getData());
$person->history->populate($request->getData());
```

**Comment by alexandernst** [comment]
@klimov-paul While I have a pro-standing point of view for code sharing, I really don't think such thing should be done here. NoSQL/MongoDB is quite complex by itself, there is no need to make it even more complex trying to share code with anything else.

Going further, the main point of Yii is simplicity (correct me if I'm wrong), and I can't see how forcing devs to create multiple sub-document models and mapping those to a single model is making anything simple. I'm afraid that this path will ultimately end up in a Symfony-like crappy structure full of models and models and more models and simplicity would be lost by far.

Also, making sub-document models is kind of awkward as they don't actually represent anything meaningful by themselves, they just mean something in the context of their parent model/document, which is why having multiple/different models/classes doesn't make much sense. Take for example my last comment. What sub-model should one create to represent an object of `history->logins` and `history->purchases`? They have almost the same fields, but that `almost` is not `the same`, which makes it impossible to handle both things with a single model, thus forcing multiple models/classes to handle such a silly (2 fields!) entry.

Furthermore, a sub-model structure as you suggest will (and it does) create the complexities you described in your `Validation` point.

To sum up my point: you're trying to represent/map a schema-less structure with well defined models/classes, which just can't be done.  The hole point of NoSQL is to store data in a schema-less way*. And that is why a model/sub-model structure (instead of plain PHP arrays/objects) will make using MongoDB weird and it will limit things.

* Take for example a "log analyzer" application that would save the columns of the log files of each application in a document, in the same collection. Apache log doesn't have the same columns compared to MySQL, or NGINX, or ... you get the point, right? That just can't be represented with a single model.

**Comment by klimov-paul** [comment]
@alexandernst, well I don't quite understand your concern about this issue then.

Just lets follow this once again:

1) Manipulation

You wish:
```php
$person = Perso::find();
$person->history->purchases->online[1]->['amount'];
```
How this is different from current state:
```php
$person = Perso::find();
$person['history']['purchases']['online'][1]['amount'];
```
Do you simply dislike brackets or what?

2) Validation

You may create your own set of validators, which will retrieve the data to be validated in your way, e.g. using dot syntaxt: `history.date`. And that be off. How does this relats to dub-documenting then?

3) Population 

Same as at 1)

And where the lazy loading should be then?

**Comment by alexandernst** [comment]
@klimov-paul I'm fine with brackets instead of arrows (even though I prefer arrows, it feels more natural and less "I'm accessing a deeply nested pile of something").
My concert is that your approach requires a huge amount of boilerplate, that is, creating a new Model/Class for each sub-document instead of defining everything in a single Model (where it actually belongs to, as that is how it's represented at the database level).

About the validation, I was talking about the validation-triggering.

**Comment by alexandernst** [comment]
@klimov-paul @Sammaye @cebe @tadaszelvys While we're discussing this trying to get to something, can we meanwhile patch Yii2 so that 

    $person->credentials['password'] = 'qwerty';

doesn't fail with

    Indirect modification of overloaded property app\models\Person::$credentials has no effect

?

The patch is really simple and I don't think it introduces any bugs.

Just convert the `__get` method in [BaseActiveRecord](https://github.com/yiisoft/yii2/blob/99a612e1a6e0beab9deced5d8242f3f25de99ef0/framework/db/BaseActiveRecord.php#L236) to `&__get`.  (add the `&`).

I can write a patch if you'll accept it.

**Comment by cebe** [comment]
looks like a simple thing but we need to figure out all the consequences. Go ahead with a pull request, we'll see how it will work.

**Comment by alexandernst** [comment]
@cebe Done https://github.com/yiisoft/yii2/pull/8002

**Comment by klimov-paul** [comment]
We have already similar discussion at #1187.
Modifying of the `__get()` in this way will make code unpredictable. You will loose any ability of actual copying of attrbite and virtual property values.
Check following code:
```php
$model->someAttribute = 'some/path';
$value = $model->someAttribute;
$value = str_replace('/', '---', $value);
var_dump($model->someAttribute); // outputs 'some---path' !!!
```
So I am afraid we can afford such change.
As it suggested at #1187 you need to get array attribute value first, modify it at then set back.

**Comment by alexandernst** [comment]
@klimov-paul Ok, I see your point, but you should see mine. It's completely unacceptable that the only way to work with MongoDB data is retrieving, modifying and then patching the documents with the modified arrays.

There must be another way which we could use to allow some simpler way to work with nested objects.

**Comment by klimov-paul** [comment]
That is why this particular issue still stands.

**Comment by alexandernst** [comment]
@klimov-paul Ok. Is your final opinion "go with models/classes for each sub-document" or are you open-minded to my idea of a single model/class for a document, no matter how deeply nested it is?

Either ways I think you should ask a few more MongoDB users, see what they think.

**Comment by Sammaye** [comment]
Passing __get by reference will break a lot of things. The PHP team have already talked about whether to solve this in the PHP core and their final thought was no since the complexity is just too great to stop indirect overload yet not pass by reference and __get was never designed like this.

Making __get pass by reference will break pretty much any of my code which does not use MongoDB subdocuments in the fashion stated here. It will also mean I need to remember to `clone` every property, including subdocuments, if I wish to manipulate them outside of the models own reference.

You want to be able to set the models subdocument property like you could if the property actually existed but you also want to be able to initiate a copy-on-write scenario like you could if you didn't pass by reference.

**Comment by alexandernst** [comment]
@Sammaye Yes, that is correct. And I do see your point (and @klimov-paul 's). `&__get` isn't the best idea as it will break things (I didn't realized that before, but @klimov-paul made a point there and I agree).

What about a class/data-holder with some tweaked `__set` and `__get` methods that can be used only with MongoDB, that would store/create keys/values as needed, in an object fashioned way? Something like http://stackoverflow.com/questions/10454779/php-indirect-modification-of-overloaded-property/10455217#10455217

That will remove the `&__get` issue.

**Comment by klimov-paul** [comment]
@alexandernst, the link you provided does not change much. It will still have same problem of referencing and copying: reduced - perhaps, but not eliminated.

> are you open-minded to my idea of a single model/class for a document, no matter how deeply nested it is?

The problem here is the `yii\base\Model` does NOT support any complex type attributes at all! It does not support this at concept level. Model in Yii is a bacth of scalar data. It handles attribute management and validation assuming those are scalar types. The model itself is a replacement of complex PHP type such as `StdClass` or associative array.
I am afraid you are asking for the not implementable feature. The model, which is a thing to store scalar data, can not be bend to store complex data. Creating an ActiveRecord for the nested documents contradicts `yii\base\Model`paradigm.
That is why things are not going well in this issue. Tha is why I placed a recommendation to avoid nested attributes if possible:
https://github.com/yiisoft/yii2-mongodb/blob/master/docs/guide/usage-embedded-documents.md

At the present state I can not see other option but usage of nested models with lazy loading.

**Comment by alexandernst** [comment]
@klimov-paul Yes, I already understood your point. Yii's base model is not designed to handle nested data. Maybe here is where we should start?

I mean, MongoDB, (probably any other NoSQL database), ElasticSearch, all those require something that Yii's base model isn't designed to do. There are two possible paths from here:

* Ignore the hole issue itself and just try to patch MongoDB/ElasticSearch until a kind-of-usable (but usability-awkward and boilerplate-bloated) state is reached
* Think about an actual re-design of base model and take an advantage of the BC in Yii 2.1

I know this is not a silly change. I know it will probably require quite a lot of work. But think it the other way: if this isn't done, in what state would MongoDB be left?

Once again, sorry for insisting on this one, but nested models really doesn't sound like a good idea (see my previous concerns) and IMHO it just tries to hide a design issue that hasn't been addressed yet.

**Comment by klimov-paul** [comment]
Refactoring of `yii\base\Model` in the way it can support complex attributes will make it very complex.
Note: this is not just the matter of getting or manipulating complex attributes. Many other things are involved here:
- form input name generation
- label generation
- working with widgets such as `GridView`
- etc.

All those aspects will be affected and overcomplicated. This does not sounds like a simple solution.

I can not decide if base model can be modified - the whole team should agree with it. Sinse such change makes sense only for particular extensions, which are not in common use, such change does not seem to be logical and does not fit to the Yii paradigm.

**Comment by alexandernst** [comment]
@klimov-paul Then the only other viable way to get MongoDB/ElasticSearch to a non-awkrad usable state is to completely separate them from Yii and re-implement BaseModel. (unless I'm missing some other option)

Can you ping the rest of the team so everybody can cast a vote and decide what is going to be?

**Comment by Sammaye** [comment]
If you treat a subdocument as you would a join then many of the complexities fail to exist. For example label generation, espcially with `somefield.somelabel`

**Comment by Sammaye** [comment]
In fact I even toyed around with the idea of getting around the __get problem by using a join:
```php
public function getRoles()
{
    return $this->hasEmbedded('roles', 'one', 'object');
}
```
or something.

**Comment by Sammaye** [comment]
Of course, the reason why I didn't continue with recoding the whole of the Model class was because it still needs to be kept upto with the rest of Yii2. So whenever a change occurs in the Yii2 model it needs to be moved over to the recoded model etc etc.

Basically I think it would be a bad idea to reimplement the model. I tried this and threw it out before I finished. Just implementing your own validator map results in enough maintanence.

**Comment by alexandernst** [comment]
@Sammaye Maintaining the base model class isn't that much work (after the initial amount of work required to make it work is done). It's not like BaseModel receives that much changes. In fact, the amount of changes done in the base of Yii are not that much, which is very normal after a project reaches certain maturity point.

**Comment by Sammaye** [comment]
It handles quite a lot, remember that the base model is extended in everything too, from yii2-mongodb to just normal old active record. You would need to extend it's recoding to all levels of Yii2. It also handles pretty much anything that is not DB/event/behaviour (yes and no on that last one) related in the ActiveRecord. 

It is quite a lot.

**Comment by Sammaye** [comment]
Yii2 would probably need to split out the files to Model and BaseModel for the extending to work properly.

**Comment by alexandernst** [comment]
@Sammaye I'm not saying the opposite. It's quite a big change and it will require quite some work. My points are

* Unless I'm missing something, it's the only way things could start working for NoSQL/ElasticSearch
* After the initial amount of work is done, only small patches should be ported from Yii2 to Yii2-NestedBaseModel (or whatever it's called)
* It will fix a design issue that currently blocks pretty much anything that can't be represented with plain SQL data structure (which is a lot).

**Comment by klimov-paul** [comment]
See my implementation at https://github.com/yiisoft/yii2-mongodb/pull/13

**Comment by klimov-paul** [comment]
Here is a extension, which covers my proposal from https://github.com/yiisoft/yii2-mongodb/pull/13.

https://github.com/yii2tech/embedded

You can use, if you want.
We may adopt this extension later, if the core team will agree.

**Comment by alexandernst** [comment]
@cebe @samdark I know @klimov-paul made an extension (see last comment), but I'm wondering what is the state of this? It has been 9 months since I pushed the conversation, but feedback stopped at some point.

Are you considering supporting (actual support, not the current state) Mongo/ElasticSearch? If so, are there any blocking issues that should be fixed before considering this one?
What about the extension? Why is not merged in core? Is it missing something?

**Comment by klimov-paul** [comment]
@alexandernst, we understand your concerns. However all members of the core team are very busy lately. Support for the embedded documents is a serious complex issue, which should not be decided in rush.
I have released an extension as a solution, which can be used in meantime.
Hpefully in near future, the core team will have more available resources, which can be spent on this issue as well.

**Comment by alexandernst** [comment]
Just pinging here to check how is the team going on this? Did you guys manage to find some time to look at this (both the extension proposed by @klimov-paul and the issue itself)? Any news?

**Comment by arash26** [comment]
@klimov-paul could I use extension on production env?

**Comment by klimov-paul** [comment]
While there is no stable version it is a bit risky.
I am planning to tag the stable release before the New Year. After that - you certainly can.

**Comment by ptheofan** [comment]
@arash26 you might also want to try https://github.com/consultnn/yii2-mongodb-embedded
It works fine. Some important features are missing (i.e. dirty attributes - which attributes have changed and such), but they are easy to implement and the way it's build gives you the freedom of doing so.

**Comment by LAV45** [comment]
:+1:  https://github.com/consultnn/yii2-mongodb-embedded

**Comment by alexandernst** [comment]
Same as last time I wrote in this issue. What is the state of this? Has yii core team reached some consensus on this? Are there any objections on using @klimov-paul 's extension?

**Comment by dynasource** [comment]
because this is an extension and not a core element, it will not have our main attention. Of course your can use the solutions that are suggested. Maybe in the future, we will give official support on this.
